stream_alert_cli/test.py

```diff
@@ -281,24 +284,73 @@ def test_rule(rule_name, test_record, formatted_record):
                 record for the service to be tested
 
         Returns:
-            [bool] boolean indicating if this rule passed
+            [list] alerts that hit for this rule
+            [integer] count of expected alerts for this rule
+            [bool] boolean where False indicates errors occurred during processing
         """
         event = {'Records': [formatted_record]}
 
         expected_alert_count = test_record.get('trigger_count')
         if not expected_alert_count:
-            expected_alert_count = (0, 1)[test_record['trigger']]
+            expected_alert_count = 1 if test_record['trigger'] else 0
+
+        # Run the rule processor. Passing mocked context object with fake
+        # values and False for suppressing sending of alerts
+        processor = StreamAlert(self.context)
+        success = processor.run(event)
 
-        # Run the rule processor. Passing 'None' for context
-        # will load a mocked object later
-        alerts = StreamAlert(None, True).run(event)
+        if not success:
+            payload = StreamPayload(raw_record=formatted_record)
+            classifier = StreamClassifier(config=load_config())
+            classifier.map_source(payload)
+            logs = classifier._log_metadata()
+            self.analyze_log_delta(logs, rule_name, test_record)
+
+        alerts = processor.get_alerts()
 
         # we only want alerts for the specific rule being tested
         alerts = [alert for alert in alerts
                   if alert['rule_name'] == rule_name]
 
         return alerts, expected_alert_count
 
+    def analyze_log_delta(self, logs, rule_name, test_record):
+        """Provide some additional context on why this test failed
+
+        Args:
+            logs [dict]: dictionary containing all of the log schema
+                information for the source/entity
+            rule_name [str]: name of rule being tested
+            test_record [dict]: actual record data being tested
+        """
+        rule_info = StreamRules._StreamRules__rules[rule_name]
```

Comment:
I'm updating this with a helper function on `StreamRules` that does this same thing. The tl;dr for why this works is here: https://stackoverflow.com/questions/797771/python-protected-attributes